modules/livestore/partition_reader.go

```diff
@@ -91,21 +102,59 @@ func (r *PartitionReader) running(ctx context.Context) error {
 
 		r.recordFetchesMetrics(fetches)
 		if offset := r.consumeFetches(ctx, fetches); offset != nil {
-			r.commitOffset(ctx, *offset)
+			r.storeOffsetForCommit(ctx, offset)
 		}
 	}
 
 	return nil
 }
 
+func (r *PartitionReader) storeOffsetForCommit(ctx context.Context, offset *kadm.Offset) {
+	if r.commitInterval == 0 { // Sync commits
+		r.commitOffset(ctx, *offset)
+	}
+
+	r.offsetWatermark.Store(offset)
+	r.hasUncommittedOffsets.Store(true)
+}
+
 func (r *PartitionReader) stop(error) error {
 	level.Info(r.logger).Log("msg", "stopping partition reader")
 
+	r.wg.Wait()
+
 	r.client.Close()
 
 	return nil
 }
 
+func (r *PartitionReader) commitLoop(ctx context.Context) {
+	defer r.wg.Done()
+
+	if r.commitInterval == 0 { // Sync commits
```

Comment:
Can't this goroutine to store a copy of the last offset? That way we won't need a second atomic bool to know if the value was updated.

Reply:
I'm not sure what you mean by this, can you elaborate?

Reply:
Sure, the `hasUncommittedOffsets atomic.Bool` is used to flag if there are uncommitted offsets or not. It could be possible to store a copy of the last watermark locally

```go
offset := r.offsetWatermark.Load()
if offset == nil || offset == lastOffset{
   return
}
```

That way, `hasUncommittedOffsets` won't be necessary